### PR TITLE
Enable optional case insensitive

### PR DESCRIPTION
**Pull request by tolemac**
In order to close this router issue: https://github.com/aurelia/router/issues/271

I have added the posibility of indicate if you want the router to be case sensitive or not, by default is true, but you can set caseSensitive to false using`setCaseSensitive` method on `RouteRecognizer` class.

### CODE REVIEW

**Comment by tolemac** [comment]
I have added the posibility of set caseSensitive property per each route.

**Comment by EisenbergEffect** [comment]
@bryanrsmith Can you review this PR and advise? Thanks!

File: src/route-recognizer.js [hunk 1]
```diff
@@ -17,6 +17,17 @@ export class RouteRecognizer {
...
     this.rootState = new State();
     this.names = {};
+    this.caseSensitive = true;
+  }
+
+  /**
+  * Set case sensitive to recognizer object. Default value is true.
+  * You can set `caseSensitive` property for each route.
+  *
+  * @param caseSensitive The route to add.
```
Comment (bryanrsmith) [inline src/route-recognizer.js]: This comment doesn't match the param

File: src/route-recognizer.js [hunk 2]
```diff
@@ -17,6 +17,17 @@ export class RouteRecognizer {
...
+  */
+  setCaseSensitive(caseSensitive : boolean) {
```
Comment (bryanrsmith) [inline src/route-recognizer.js]: No space before `:`, please.

File: src/route-recognizer.js [hunk 3]
```diff
@@ -37,6 +48,7 @@ export class RouteRecognizer {
     let routeName = route.handler.name;
     let isEmpty = true;
     let segments = parse(route.path, names, types);
+    const caseSensitiveRoute = route.caseSensitive === undefined ? this.caseSensitive : route.caseSensitive;
```
Comment (bryanrsmith) [inline src/route-recognizer.js]: I think it'd be better to check if `route.caseSensitive` is a boolean, to guard against invalid input.

File: src/route-recognizer.js [hunk 4]
```diff
@@ -73,7 +85,7 @@ export class RouteRecognizer {
     }
 
     currentState.handlers = handlers;
-    currentState.regex = new RegExp(regex + '$');
+    currentState.regex = caseSensitiveRoute ? new RegExp(regex + '$') : new RegExp(regex + '$', 'i');
```
Comment (bryanrsmith) [inline src/route-recognizer.js]: It'd be great to avoid duplicating the ctor call, and simply switch the second parameter out instead.

File: src/route-recognizer.js [hunk 5]
```diff
@@ -334,9 +346,12 @@ function findHandler(state, path, queryParams) {
   return result;
 }
 
-function addSegment(currentState, segment) {
+function addSegment(currentState, segment, caseSensitive) {
   let state = currentState;
   segment.eachChar(ch => {
+    if (!caseSensitive && segment instanceof StaticSegment) {
+      ch.validChars = ch.validChars.toUpperCase() + ch.validChars.toLowerCase();
+    }
```
Comment (bryanrsmith) [inline src/route-recognizer.js]: It'd be cleaner for the `StaticSegment` ctor to receive the `caseSensitive` value and adjust its own `validChars`.

File: test/route-recognizer.spec.js [hunk 6]
```diff
@@ -1,53 +1,53 @@
...
+  isDynamic: true,
+  path: '/star/test/path',
+  params: { path: 'test/path' }
+}, {
+  title: 'dynamic star routes',
+  route: { 'path': 'dynamic/:id/star/*path', 'handler': { 'name': 'star' } },
+  isDynamic: true,
+  path: '/dynamic/foo/star/test/path',
+  params: { id: 'foo', path: 'test/path' }
+}];
```
Comment (bryanrsmith) [inline test/route-recognizer.spec.js]: Please make whitespace changes in a separate PR so it's easier to see the functional changes made.

**Comment by EisenbergEffect** [comment]
@tolemac Thanks for contributing. @bryanrsmith has provided you with some feedback. If you can address the issues he has pointed out, then we can probably get this merged in. Please ping me @EisenbergEffect and @bryanrsmith after you update the code. Thanks!

**Comment by tolemac** [comment]
I have read the contributing guidelines and I can see I haven't made the things in the correct way.
Let me close this pull request and do the things better.
